nixos/modules/services/matrix/synapse.nix

```diff
@@ -1022,15 +1098,35 @@ in {
           This is done by default unless you manually configure either of those settings.
         '';
       }
-    ];
+    ]
+    ++ (map (listener: {
+      assertion = (listener.path == null) != (listener.bind_addresses == null);
+      message = ''
+        Listeners require either a UNIX domain socket `path` or `bind_addresses` for a TCP socket.
+      '';
+    }) cfg.settings.listeners)
+    ++ (map (listener: {
+      assertion = listener.path != null -> (listener.bind_addresses == null || listener.port == null || listener.tls == null);
```

Comment:
I think this needs to be `listener.path != null -> (listener.bind_addresses == null && ... && ...)`.
Right now, the assertion is still `true` if `path != null` and `port == null`, but `bind_addresses` isn't null.